### ISSUE TITLE
[Bug] GDN prefill kernel produces NaN

### ISSUE THREAD

**Issue by ZJY0516** [flashinfer-ai/flashinfer#2490]
[gdn_debug_inputs.pt.zip](https://github.com/user-attachments/files/25076044/gdn_debug_inputs.pt.zip)

I'm trying to use flashinfer gdn prefill kernel in vLLM, see https://github.com/vllm-project/vllm/pull/32846.

But I find it will always produce NaN

```python
import torch
from flashinfer.gdn_prefill import chunk_gated_delta_rule


def main():
    data = torch.load("gdn_debug_inputs.pt", map_location="cuda")

    q = data["q"]
    k = data["k"]
    v = data["v"]
    g = data["g"]
    beta = data["beta"]
    state = data["state"]
    cu_seqlens = data["cu_seqlens"]

    output, final_state = chunk_gated_delta_rule(
        q=q,
        k=k,
        v=v,
        g=g,
        beta=beta,
        initial_state=state,
        output_final_state=True,
        cu_seqlens=cu_seqlens,
    )
    print(output, final_state)


if __name__ == "__main__":
    main()

```

<details>
<summary>Output</summary>

```text
tensor([[[nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         ...,
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan]],

        [[nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         ...,
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan]],

        [[nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         ...,
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan]],

        ...,

        [[nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         ...,
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan]],

        [[nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         ...,
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan]],

        [[nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         ...,
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan],
         [nan, nan, nan,  ..., nan, nan, nan]]], device='cuda:0',
       dtype=torch.bfloat16) tensor([[[[nan, nan, nan,  ..., nan, nan, nan],
          [nan, nan, nan,  ..., nan, nan, nan],
          [nan, nan, nan,  ..., nan, nan, nan],
          ...,
          [nan, nan, nan,  ..., nan, nan, nan],
          [nan, nan, nan,  ..., nan, nan, nan],
          [nan, nan, nan,  ..., nan, nan, nan]],

         [[nan, nan, nan,  ..., nan, nan, nan],
          [nan, nan, nan,  ..., nan, nan, nan],
          [nan, nan, nan,  ..., nan, nan, nan],
          ...,
          [nan, nan, nan,  ..., nan, nan, nan],
          [nan, nan, nan,  ..., nan, nan, nan],
          [nan, nan, nan,  ..., nan, nan, nan]],

         [[nan, nan, nan,  ..., nan, nan, nan],
          [nan, nan, nan,  ..., nan, nan, nan],
          [nan, nan, nan,  ..., nan, nan, nan],
          ...,
          [nan, nan, nan,  ..., nan, nan, nan],
          [nan, nan, nan,  ..., nan, nan, nan],
          [nan, nan, nan,  ..., nan, nan, nan]],

         ...,

         [[nan, nan, nan,  ..., nan, nan, nan],
          [nan, nan, nan,  ..., nan, nan, nan],
          [nan, nan, nan,  ..., nan, nan, nan],
          ...,
          [nan, nan, nan,  ..., nan, nan, nan],
          [nan, nan, nan,  ..., nan, nan, nan],
          [nan, nan, nan,  ..., nan, nan, nan]],

         [[nan, nan, nan,  ..., nan, nan, nan],
          [nan, nan, nan,  ..., nan, nan, nan],
          [nan, nan, nan,  ..., nan, nan, nan],
          ...,
          [nan, nan, nan,  ..., nan, nan, nan],
          [nan, nan, nan,  ..., nan, nan, nan],
          [nan, nan, nan,  ..., nan, nan, nan]],

         [[nan, nan, nan,  ..., nan, nan, nan],
          [nan, nan, nan,  ..., nan, nan, nan],
          [nan, nan, nan,  ..., nan, nan, nan],
          ...,
          [nan, nan, nan,  ..., nan, nan, nan],
          [nan, nan, nan,  ..., nan, nan, nan],
          [nan, nan, nan,  ..., nan, nan, nan]]]], device='cuda:0')
```

</details>

**Comment by yzh119** [comment]
Hi @ZJY0516 there is a layout mismatch: the flashinfer's state layout is `[N, H, V, K]` instead of `[N, H, K, V]` (which vllm originally expects), because K-last layout is more friendly to kernels to get best performance, not sure if it's the reason for the NaN output.

**Comment by ZJY0516** [comment]
> Hi [@ZJY0516](https://github.com/ZJY0516) there is a layout mismatch: the flashinfer's state layout is `[N, H, V, K]` instead of `[N, H, K, V]` (which vllm originally expects), because K-last layout is more friendly to kernels to get best performance, not sure if it's the reason for the NaN output.

Yes, but vLLM has already changed to `[N, H, V, K]`, see https://github.com/vllm-project/vllm/pull/33291

**Comment by guangyunh-nv** [comment]
Hi, Flash Infer GDN kernel author here.

The root cause seems to come from the "gating" space.
<img width="886" height="938" alt="Image" src="https://github.com/user-attachments/assets/bb0cbdd4-a44b-4a35-a527-92597a4c1fff" />

We have fused everthing, that is, the log and cumsum are all implemented in a single kernel. FYI, see https://github.com/flashinfer-ai/flashinfer/blob/cdbb2c3c73598bbb04a0e2dc2d7cf6ee9e5aeda5/include/flashinfer/flat/hopper/collective/flat_collective_tma_warpspecialized_delta_rule.hpp#L289-L339


```py
    output, final_state = chunk_gated_delta_rule(
        q=q,
        k=k,
        v=v,
        g=torch.exp(g),  # should fix your problem
        beta=beta,
        initial_state=state,
        output_final_state=True,
        cu_seqlens=cu_seqlens,
    )
```

**Comment by guangyunh-nv** [comment]
@yzh119  I think a proper solution to this is to add an additional flag to disambiguate the g spaces `g_space="linear" | "ln" | "log2"`. Because g in linear space and g in log space are both valid gate. log space g exists only because we want to implement numerical stable cumulative product.

**Comment by ZJY0516** [comment]
@guangyunh-nv Thank you for the excellent work and explanation

**Comment by vadiklyutiy** [comment]
> [@yzh119](https://github.com/yzh119) I think a proper solution to this is to add an additional flag to disambiguate the g spaces `g_space="linear" | "ln" | "log2"`. Because g in linear space and g in log space are both valid gate. log space g exists only because we want to implement numerical stable cumulative product.

To double check. Are you going to implement it?

**Comment by vadiklyutiy** [comment]
> > [@yzh119](https://github.com/yzh119) I think a proper solution to this is to add an additional flag to disambiguate the g spaces `g_space="linear" | "ln" | "log2"`. Because g in linear space and g in log space are both valid gate. log space g exists only because we want to implement numerical stable cumulative product.
> 
> To double check. Are you going to implement it?

@dianzhangchen would be really good if interfaces of Hopper and Blackwell implementation be identical

**Comment by guangyunh-nv** [comment]
@vadiklyutiy Yes, I can implement it.

**Comment by dianzhangchen** [comment]
@vadiklyutiy Hi, thanks for letting me know. In the current Blackwell implementation, I use the same input as FLA, which assumes g is in log space. Should I change this, or should we all use the same input as FLA's implementation?

**Comment by ZJY0516** [comment]
Do we have any benchmark results indicating which option has better accuracy?

**Comment by ZJY0516** [comment]
> [@yzh119](https://github.com/yzh119) I think a proper solution to this is to add an additional flag to disambiguate the g spaces `g_space="linear" | "ln" | "log2"`. Because g in linear space and g in log space are both valid gate. log space g exists only because we want to implement numerical stable cumulative product.

At least we should add some commnets in doc string

**Comment by guangyunh-nv** [comment]
@ZJY0516  The gating space will not affect the accuracy at all (largely). Internally, you will always do the linear space gating cumprod in log2 space via cumsum. Conversion between them is trivial.

- In `linear` space: load `g_linear`, take `log2`, `warp shfl` to compute cumsum. `exp2` to get back in linear space.
- In `log2` space: load `g_log2`,  `warp shfl` to compute cumsum. `exp2` to get back in linear space.
- In `ln` space, load `g_ln`, div `LN2` constant to get `g_log2`, `warp shfl` to compute cumsum. `exp2` to get back in linear space.

Do it in log2 space is not mandatory. But the hardware do not actually support e base computation directly, it always changes the base for you.

**Comment by zhiyuan1i** [comment]
FLA has a plan to change the layout of the state to [N,H,V,K] I would like to ask if we need to keep the prefill kernel interface and definition of gdn consistent with FLA to reduce the mental burden of maintenance? And FLA now supports op dispatch(really simple and easy dispatch, designed for training https://github.com/fla-org/flash-linear-attention/commit/4fdb39a266e2937caa120f03aba62f3cb1debb57)

**Comment by yzh119** [comment]
Hi @zhiyuan1i sure it's better if we can keep layout aligned, can you let us know what's the currently gap between the prefill api's interface with FLA's?

@claude can we add an option to specify the space of g as suggested by @guangyunh-nv ? Not necessarily using the name `g_space` but we should explicitly articulating the behavior.

**Comment by vadiklyutiy** [comment]
Did Claude make what was expected? Do we need just simply call `exp`? 
As I understand from descr above internally kernel make a `ln`. Is it correct understanding? May we avoid doing `ln(exp())`?

**Comment by guangyunh-nv** [comment]
Yes, it is correct.

**Comment by ZJY0516** [comment]
flashinfer reuses FLA interface for prefill: `chunk_gated_delta_rule`, but use different interface for decoding: `gated_delta_rule_decode_pretranspose, gated_delta_rule_decode, gated_delta_rule_mtp,`  whereas FLA uses `fused_recurrent_gated_delta_rule`. Are we considering aligning these interfaces as well? cc @yzh119 @zhiyuan1i
